data-prepper-plugins/s3-source/src/main/java/org/opensearch/dataprepper/plugins/source/ScanObjectWorker.java

```diff
@@ -0,0 +1,186 @@
+            final S3ObjectReference s3ObjectReference = S3ObjectReference.bucketAndKey(scanOptionsBuilder.getBucket(), key).build();
+            processS3ObjectKeys(s3ObjectReference,s3ObjectHandler, scanOptionsBuilder);
+        });
+    }
+
+
+    private void processS3ObjectKeys(final S3ObjectReference s3ObjectReference, final S3ObjectHandler s3ObjectHandler, final ScanOptionsBuilder scanOptionsBuilder){
+        final S3ObjectDetails s3ObjDetails = getS3ObjectKeyLastModifiedTime(s3ObjectReference);
+        final boolean isKeyMatchedBetweenTimeRange = isKeyMatchedBetweenTimeRange(s3ObjDetails.getS3ObjectLastModifiedTimestamp(),
+                scanOptionsBuilder.getRange(),
+                LocalDateTime.parse(scanOptionsBuilder.getStartDate()));
+        if(isKeyMatchedBetweenTimeRange && (isKeyProcessedByS3Scan(s3ObjDetails))){
+            updateKeyProcessedByS3Scan(s3ObjDetails);
+            try{
+                s3ObjectHandler.parseS3Object(s3ObjectReference);
+            }catch (IOException ex){
+                deleteKeyProcessedByS3Scan(s3ObjDetails);
+                LOG.error("Error while process the parseS3Object. ",ex);
+            }
+        }
+    }
+    /**
+     * Method will identify already processed key.
+     * @return boolean
+     */
+    public boolean isKeyProcessedByS3Scan(final S3ObjectDetails s3ObjectDetails) {
+        final S3ObjectDetails mapDbs3ObjectDetails = stateSaveMap.get((s3ObjectDetails.getBucket()+s3ObjectDetails.getKey()));
+        return mapDbs3ObjectDetails == null;
+    }
+
+    /**
+     * store the processed bucket key in the map.
+     */
+    public void updateKeyProcessedByS3Scan(final S3ObjectDetails s3ObjectDetails) {
+        stateSaveMap.put((s3ObjectDetails.getBucket() + s3ObjectDetails.getKey()),s3ObjectDetails);
+    }
+    private void deleteKeyProcessedByS3Scan(S3ObjectDetails s3ObjDetails) {
+        stateSaveMap.remove(s3ObjDetails.getBucket() + s3ObjDetails.getKey());
+    }
+
+    /**
+     * fetch the s3 object last modified time.
+     * @return S3ObjectDetails
+     */
+    public S3ObjectDetails getS3ObjectKeyLastModifiedTime(final S3ObjectReference s3ObjectReference){
+        GetObjectRequest getObjectRequest = GetObjectRequest.builder().bucket(s3ObjectReference.getBucketName()).key(s3ObjectReference.getKey()).build();
+        ResponseInputStream<GetObjectResponse> s3ObjectResp = s3Client.getObject(getObjectRequest);
+        final Instant instant = s3ObjectResp.response().lastModified();
+        ZonedDateTime zonedDateTime = instant.atZone(ZoneId.systemDefault());
+        return new S3ObjectDetails(s3ObjectReference.getBucketName(),s3ObjectReference.getKey(),zonedDateTime.toLocalDateTime());
+    }
+
+    /**
+     * used for identifying the slurping start date based on the range string.
+     * @return LocalDateTime
+     */
+    public LocalDateTime getSlurpingStartDateByRange(final String rangeString,final LocalDateTime startDateTime){
```

Comment:
I think you have to use `DataPrepperDurationDeserializer` instead of writing parsing code.

Reply:
Thanks for your review krishna, As discussed, Can you please get back on us DataPrepperDurationDeserializer capability.